### PR TITLE
[GeoMechanicsApplication] Unit tests for TransientPwElement

### LINKED ISSUE #12996: [GeoMechanicsApplication] Assessment of making TransientPwElement inherent from Element
Currently, TransientPwElement is inherent from UPw element which leads to unnecessary methods.  At the same time TransientPwLineElement is inherent from Element and it has much smaller number of methods.

This issue goal is to investigate making TransientPwElement inherent from Element.

### PR DESCRIPTION
**Pull request by markelov208**
**📝 Description**
A brief description of the PR.

Unit tests have been created for the following TransientPwElement functions
1. Create
2. Initialize
3. Check
4. InitializeSolutionStep
5. InitializeNonLinearIteration
6. FinalizeNonLinearIteration
7. FinalizeSolutionStep
8. three CalculateOnIntegrationPoints
Two unit tests (2D and 3D elements) call CalculateLocalSystem that calls
9. CalculateAll
10. InitializeElementVariables
11. CalculateAndAddLHS
12. CalculateAndAddCompressibilityMatrix
13. CalculateAndAddRHS
14. CalculateAndAddPermeabilityFlow
15. CalculateAndAddFluidBodyFlow
16. CalculateAndAddCompressibilityFlow
17. CalculateKinematics

An extra unit test is created for functions that return zero values:
18. CalculateMassMatrix
19. CalculateDampingMatrix
20. GetValuesVector
21. GetFirstDerivativesVector
22. GetSecondDerivativesVector

The following function have an empty body:
23. InitializeNonLinearIteration
24. FinalizeNonLinearIteration

### CODE REVIEW

File: applications/GeoMechanicsApplication/custom_elements/transient_Pw_element.hpp [hunk 1]
```diff
@@ -19,7 +19,6 @@
 #include "custom_elements/U_Pw_small_strain_element.hpp"
 #include "custom_utilities/element_utilities.hpp"
 #include "custom_utilities/stress_strain_utilities.h"
-#include "geo_mechanics_application_variables.h"
```
Comment (rfaasse) [inline applications/GeoMechanicsApplication/custom_elements/transient_Pw_element.hpp:22]: Nice catch! Removing these when redundant will speed up compilation times (especially large headers like this one)

File: applications/GeoMechanicsApplication/tests/cpp_tests/custom_elements/test_transient_Pw_element.cpp [hunk 2]
```diff
@@ -0,0 +1,696 @@
...
+    // Act
+    const auto              dummy_process_info = ProcessInfo{};
+    Element::DofsVectorType degrees_of_freedom;
+    p_element->GetDofList(degrees_of_freedom, dummy_process_info);
+
+    // Assert
+    KRATOS_EXPECT_EQ(degrees_of_freedom.size(), 3);
+    for (auto p_dof : degrees_of_freedom) {
+        KRATOS_EXPECT_EQ(p_dof->GetVariable(), WATER_PRESSURE);
+    }
```
Comment (rfaasse) [inline applications/GeoMechanicsApplication/tests/cpp_tests/custom_elements/test_transient_Pw_element.cpp]: I think we could use an algorithm here (might be syntax errors here, but something like this):
```suggestion
    KRATOS_EXPECT_TRUE(std::all_of(degrees_of_freedom.begin(), degrees_of_freedom.end(), [](auto pdof){return p_dof->GetVariable() == WATER_PRESSURE;});
```

File: applications/GeoMechanicsApplication/tests/cpp_tests/custom_elements/test_transient_Pw_element.cpp [hunk 3]
```diff
@@ -0,0 +1,696 @@
...
+    KRATOS_EXPECT_EQ(r_constitutive_law_vector.size(), number_of_integration_points);
+    for (const auto& constitutive_law : r_constitutive_law_vector) {
+        KRATOS_EXPECT_EQ(constitutive_law, nullptr);
+    }
+
+    const auto& r_retention_law_vector = element.mRetentionLawVector;
+    KRATOS_EXPECT_EQ(r_retention_law_vector.size(), number_of_integration_points);
+    for (const auto& retention_law : r_retention_law_vector) {
+        KRATOS_EXPECT_NE(dynamic_cast<SaturatedLaw*>(retention_law.get()), nullptr);
+    }
```
Comment (rfaasse) [inline applications/GeoMechanicsApplication/tests/cpp_tests/custom_elements/test_transient_Pw_element.cpp]: Similar to my other comment, we should be able to use `std::none_of` here (KRATOS_EXPECT_TRUE(std::none_of... etc)

File: applications/GeoMechanicsApplication/tests/cpp_tests/custom_elements/test_transient_Pw_element.cpp [hunk 4]
```diff
@@ -0,0 +1,696 @@
...
+    auto p_element = CreateTransientPwElementWithPWDofs<2, 3>(r_model_part, std::make_shared<Properties>());
+    const auto dummy_process_info = ProcessInfo{};
+
+    // Act
+    p_element->InitializeSolutionStep(dummy_process_info);
+
+    // Assert
+    for (auto& node : p_element->GetGeometry()) {
+        KRATOS_EXPECT_EQ(node.FastGetSolutionStepValue(HYDRAULIC_DISCHARGE), 0.0);
+    }
```
Comment (rfaasse) [inline applications/GeoMechanicsApplication/tests/cpp_tests/custom_elements/test_transient_Pw_element.cpp]: We can use a std::all_of here (and also in some other places in the next tests 👍 )

**Review by rfaasse — COMMENTED** [review]
Thank you for the (tedious) work you put int adding unit tests and increasing our coverage! I just have some ideas for using algorithms, but this is very valuable!

File: applications/GeoMechanicsApplication/tests/cpp_tests/custom_elements/test_transient_Pw_element.cpp (hunk 2 above)
Comment (markelov208) [reply applications/GeoMechanicsApplication/tests/cpp_tests/custom_elements/test_transient_Pw_element.cpp]: done

File: applications/GeoMechanicsApplication/tests/cpp_tests/custom_elements/test_transient_Pw_element.cpp (hunk 3 above)
Comment (markelov208) [reply applications/GeoMechanicsApplication/tests/cpp_tests/custom_elements/test_transient_Pw_element.cpp]: done

File: applications/GeoMechanicsApplication/tests/cpp_tests/custom_elements/test_transient_Pw_element.cpp (hunk 4 above)
Comment (markelov208) [reply applications/GeoMechanicsApplication/tests/cpp_tests/custom_elements/test_transient_Pw_element.cpp]: done

**Review by markelov208 — COMMENTED** [review]
Hi Richard and Wijtze-Pieter, thank you very much for making the code better. I added on more unit test and remove a few lines from the code. Hope the coverage will be higher now.

File: applications/GeoMechanicsApplication/tests/cpp_tests/custom_elements/test_transient_Pw_element.cpp [hunk 5]
```diff
@@ -693,4 +695,59 @@ KRATOS_TEST_CASE_IN_SUITE(TransientPwElement3D4N_CalculateLocalSystem, KratosGeo
...
+    // Act
+    element.GetSecondDerivativesVector(actual_vector, dummy_process_info);
+
+    // Assert
+    KRATOS_EXPECT_EQ(actual_vector.size(), n_DoF);
+    KRATOS_EXPECT_VECTOR_EQ(actual_vector, expected_vector);
+
+    // Act empty functions to increase the coverage
+    element.InitializeNonLinearIteration(dummy_process_info);
+    element.FinalizeNonLinearIteration(dummy_process_info);
```
Comment (rfaasse) [inline applications/GeoMechanicsApplication/tests/cpp_tests/custom_elements/test_transient_Pw_element.cpp]: I'm not sure if we should add these lines, since nothing is asserted (of course right now there's nothing to be asserted). However, if we ever add any functionality to these functions, it seems like they're covered (because of this test), but actually they're only called.

File: applications/GeoMechanicsApplication/tests/cpp_tests/custom_processes/test_apply_c_phi_reduction_process.cpp [hunk 6]
```diff
@@ -187,8 +187,8 @@ KRATOS_TEST_CASE_IN_SUITE(CheckFailureEmptyModelPartApplyCPhiReductionProcess, K
 
 KRATOS_TEST_CASE_IN_SUITE(CheckReturnsZeroForValidModelPartApplyCPhiReductionProcess, KratosGeoMechanicsFastSuite)
 {
-    Model model;
-    auto& r_model_part = PrepareCPhiTestModelPart(model);
+    Model                     model;
```
Comment (rfaasse) [inline applications/GeoMechanicsApplication/tests/cpp_tests/custom_processes/test_apply_c_phi_reduction_process.cpp:190]: Thanks for formatting these (where I forgot to)!

**Review by rfaasse — DISMISSED** [review]
Good to go in my opinion, I have one discussion point left (about calling some now empty functions and not asserting anything for the coverage), but I don't think it is blocking

**Review by rfaasse — APPROVED** [review]
In my opinion, this is good to go!
